File: sdk/android/morph-markdown/src/main/java/com/morph/markdown/RenderDebounceState.kt
package com.morph.markdown

internal sealed class RenderDebounceDecision {
	data class Schedule(val delayMs: Long) : RenderDebounceDecision()
	data class RenderNow(val autoScroll: Boolean) : RenderDebounceDecision()
	object None : RenderDebounceDecision()
}

internal class RenderDebounceState {
	private var scheduled = false
	private var pendingAutoScroll = false

	fun onAppend(final: Boolean, autoScroll: Boolean, debounceMs: Long): RenderDebounceDecision {
		pendingAutoScroll = pendingAutoScroll || autoScroll
		if (final || debounceMs <= 0L) {
			return RenderDebounceDecision.RenderNow(consumeAutoScroll())
		}
		if (scheduled) return RenderDebounceDecision.None
		scheduled = true
		return RenderDebounceDecision.Schedule(debounceMs)
	}

	fun onScheduledRender(): Boolean {
		scheduled = false
		return consumeAutoScroll()
	}

	fun cancel() {
		scheduled = false
		pendingAutoScroll = false
	}

	private fun consumeAutoScroll(): Boolean {
		val autoScroll = pendingAutoScroll
		pendingAutoScroll = false
		scheduled = false
		return autoScroll
	}
}

File: sdk/ios/MorphMarkdown/Sources/MorphMarkdown/RenderDebounceState.swift
import Foundation

enum RenderDebounceDecision: Equatable {
	case none
	case schedule(Int)
	case renderNow(Bool)
}

final class RenderDebounceState {
	private var scheduled = false
	private var pendingAutoScroll = false

	func onAppend(final: Bool, autoScroll: Bool, debounceMilliseconds: Int) -> RenderDebounceDecision {
		pendingAutoScroll = pendingAutoScroll || autoScroll
		if final || debounceMilliseconds <= 0 {
			return .renderNow(consumeAutoScroll())
		}
		guard !scheduled else { return .none }
		scheduled = true
		return .schedule(debounceMilliseconds)
	}

	func onScheduledRender() -> Bool {
		scheduled = false
		return consumeAutoScroll()
	}

	func cancel() {
		scheduled = false
		pendingAutoScroll = false
	}

	private func consumeAutoScroll() -> Bool {
		let value = pendingAutoScroll
		pendingAutoScroll = false
		scheduled = false
		return value
	}
}
